### PR TITLE
Enable firmware based device reset

### PR DESCRIPTION
**Pull request by HWXLR8**
No need to press the button anymore :) Gave this special treatment within the FW since it does not really belong with the "configuration". Let me know your thoughts!

Closes  https://github.com/HWXLR8/beef-board/issues/30

### CODE REVIEW

**Comment by ASleepyCat** [comment]
An SDVX controller only has seven buttons, so either we'd have to use a different button combo just for SDVX firmware or you'd have to change it to accommodate.

I'm also not comfortable with being able to enter into DFU potentially during gameplay, I prefer arcin's approach of only allowing it on device start up. That way we also could simplify the button combo check since we could instantly enter into DFU with an easier button combo.

**Comment by HWXLR8** [comment]
I agree that I do not want the user to accidentally enter DFU mode. How does the arcin work? Button combination on boot?

**Comment by ASleepyCat** [comment]
> I agree that I do not want the user to accidentally enter DFU mode. How does the arcin work? Button combination on boot?

[Yeah, checks B1 and B2 I think (among other things)](https://github.com/kinetic-flow/arcin-infinitas/blob/63faaff19195c45b9033c1b530c3852ee4895480/bootloader/main.cpp#L255-L275).

**Comment by HWXLR8** [comment]
I like this idea. I think holding start on boot should be sufficient. Also "start" is agnostic between controllers.

**Comment by ASleepyCat** [comment]
> I like this idea. I think holding start on boot should be sufficient. Also "start" is agnostic between controllers.

Only issue is that the start buttons don't map to the same pin (B8 (or B11?) on IIDX, B7 on SDVX, ??? on popn).

**Comment by HWXLR8** [comment]
shit, you're right. Any suggestions?

**Comment by ASleepyCat** [comment]
Why not just copy what arcin does?

**Comment by HWXLR8** [comment]
That is certainly a possibility, but I like to put some thought into my design choices before copying what someone else has done. 

One advantage of being able to enter DFU at any time is that we can develop FW without the need to touch the USB cable at all. I don't know about you, but having to unplug and replug the cable can get pretty annoying especially when making frequent small changes.

**Comment by ASleepyCat** [comment]
I don't really find it annoying having to replug the board after flashing. I'd rather do that than accidentally enter DFU tbh. You'd still have to replug anyway to get out of DFU, unless if you've found a solution for that.

We can always change it afterwards if it actually becomes too much of a hassle.

**Comment by HWXLR8** [comment]
I am working on a solution for the post-flash reset. But in the meantime I am willing to accept your suggestion. Will make modifications tmr.

**Comment by ASleepyCat** [comment]
We could make always-on DFU entry a compile flag toggle, with it being default off.

**Comment by HWXLR8** [comment]
> We could make always-on DFU entry a compile flag toggle, with it being default off.

I am cool with the current behaviour. Changes implemented as requested. Thanks for the feedback!

File: fw/beef.cpp [hunk 1]
```diff
@@ -67,11 +67,22 @@ void debounce(DebounceState* debounce, uint16_t mask) {
     (debounce->debounce(button_state & mask));
 }
 
+void check_for_dfu() {
+  button_state = 0;
```
Comment (ASleepyCat) [inline fw/beef.cpp:71]: `button_state` should be initialised to 0 so this isn't necessary.

File: fw/beef.cpp [hunk 2]
```diff
@@ -67,11 +67,22 @@ void debounce(DebounceState* debounce, uint16_t mask) {
...
+  for (int i = 0; i < BUTTONS; ++i) {
+    process_button(buttons[i].INPUT_PORT.PIN,
+                   i,
+                   buttons[i].input_pin);
+  }
+  if ((button_state & BUTTON_1) &&
+      (button_state & BUTTON_2)) {
```
Comment (ASleepyCat) [inline fw/beef.cpp:78]: This can be simplified to `button_state == (BUTTON_1 | BUTTON_2)`, or call `is_pressed(BUTTON_1 | BUTTON_2)`.

**Review by ASleepyCat — CHANGES_REQUESTED** [review]

**Comment by HWXLR8** [comment]
Made changes as requested.

**Pull request merged**